tests/poly1305.cc

```diff
@@ -0,0 +1,332 @@
+/*
+ *    Copyright 2022 Cryspen Sarl
+ *
+ *    Licensed under the Apache License, Version 2.0 or MIT.
+ *    - http://www.apache.org/licenses/LICENSE-2.0
+ *    - http://opensource.org/licenses/MIT
+ */
+
+#include <fstream>
+
+#include <gtest/gtest.h>
+#include <nlohmann/json.hpp>
+
+#include "hacl-cpu-features.h"
+#include "util.h"
+
+#include "EverCrypt_Poly1305.h"
+#include "Hacl_Poly1305_32.h"
+#include "Hacl_Streaming_Poly1305_32.h"
+
+#ifdef HACL_CAN_COMPILE_VEC128
+#include "Hacl_Poly1305_128.h"
+#include "Hacl_Streaming_Poly1305_128.h"
+#endif
+#ifdef HACL_CAN_COMPILE_VEC256
+#include "Hacl_Poly1305_256.h"
+#include "Hacl_Streaming_Poly1305_256.h"
+#endif
+
+using json = nlohmann::json;
+
+typedef struct
+{
+  std::string comment;
+  bytes key;
+  bytes text;
+  bytes tag;
+} TestCase;
+
+// -----------------------------------------------------------------------------
+
+const uint32_t POLY1305_KEY_SIZE = 32;
+const uint32_t POLY1305_TAG_SIZE = 16;
+
+void
+poly1305_mac(bytes key, bytes text, bytes& tag)
+{
+  std::cout << "Poly1305.Mac" << std::endl;
+
+  // This works everywhere. Let's use it as a base for comparisons.
+  bytes base_tag = std::vector<uint8_t>(POLY1305_TAG_SIZE);
+  Hacl_Poly1305_32_poly1305_mac(
+    base_tag.data(), text.size(), text.data(), key.data());
+
+#ifdef HACL_CAN_COMPILE_VEC128
+  if (hacl_vec128_support()) {
+    std::cout << "Poly1305.Mac (VEC128)" << std::endl;
+
+    bytes tag = std::vector<uint8_t>(POLY1305_TAG_SIZE);
+    Hacl_Poly1305_128_poly1305_mac(
+      tag.data(), text.size(), text.data(), key.data());
+
+    EXPECT_EQ(base_tag, tag)
+      << "Detected difference between base and _128 version";
+  } else {
+    std::cout << "No support for VEC128 on this CPU." << std::endl;
+  }
+#endif
+
+#ifdef HACL_CAN_COMPILE_VEC256
+  if (hacl_vec256_support()) {
+    std::cout << "Poly1305.Mac (VEC256)" << std::endl;
+
+    bytes tag = std::vector<uint8_t>(POLY1305_TAG_SIZE);
+    Hacl_Poly1305_256_poly1305_mac(
+      tag.data(), text.size(), text.data(), key.data());
+
+    EXPECT_EQ(base_tag, tag)
+      << "Detected difference between base and _256 version";
+  } else {
+    std::cout << "No support for VEC256 on this CPU." << std::endl;
+  }
+#endif
+
+  // EverCrypt
+  {
+    std::cout << "Poly1305.Mac (EverCrypt)" << std::endl;
+
+    EverCrypt_AutoConfig2_init();
+
+    bytes tag = bytes(POLY1305_TAG_SIZE);
+
+    EverCrypt_Poly1305_poly1305(
```

Comment:
So here we only get one branch of course because evercrypt picks the best option.
I think that's fine for now but please note this down somewhere to ensure it's on the issue tracker/board.

Reply:
That is one reason I wanted to have a separate issue for `EverCrypt_`. 

Should we use ...

```
  1 void EverCrypt_AutoConfig2_disable_adx();
  2 void EverCrypt_AutoConfig2_disable_aesni();
  3 void EverCrypt_AutoConfig2_disable_avx();   
  4 void EverCrypt_AutoConfig2_disable_avx512();
  5 void EverCrypt_AutoConfig2_disable_bcrypt();   
  6 void EverCrypt_AutoConfig2_disable_hacl();  
  7 void EverCrypt_AutoConfig2_disable_movbe(); 
  8 void EverCrypt_AutoConfig2_disable_openssl();
  9 void EverCrypt_AutoConfig2_disable_pclmulqdq();
 10 void EverCrypt_AutoConfig2_disable_rdrand();   
 11 void EverCrypt_AutoConfig2_disable_shaext();   
 12 void EverCrypt_AutoConfig2_disable_sse();      
 13 void EverCrypt_AutoConfig2_disable_vale();  
```

... to exercise EverCrypt a bit?